lib/linked_list.rb

```diff
@@ -12,152 +12,242 @@ def initialize(value, next_node = nil)
-    # Time Complexity:  
-    # Space Complexity
-    def find_min
-      raise NotImplementedError
+  # method to return the min value in the linked list
+  # returns the data value and not the node
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def find_min
+    return nil if @head.nil?
+
+    return nil if @head.nil?
+    current = @head
+    min = current.data
+
+    until current.next.nil?
+      current = current.next
+      if min > current.data
+        min = current.data
+      end
     end
+    return min
+  end
 
+  # method that returns the length of the singly linked list
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def length
+    current = @head
+    return 0 if @head.nil?
+    return 1 if current.next.nil?
 
-    # method that returns the length of the singly linked list
-    # Time Complexity:  
-    # Space Complexity
-    def length
-      raise NotImplementedError
+    count = 1
+    until current.next.nil?
+      current = current.next
+      count += 1
     end
+    return count
+  end
 
-    # method that returns the value at a given index in the linked list
-    # index count starts at 0
-    # returns nil if there are fewer nodes in the linked list than the index value
-    # Time Complexity:  
-    # Space Complexity
-    def get_at_index(index)
-      raise NotImplementedError
-    end
+  # method that returns the value at a given index in the linked list
+  # index count starts at 0
+  # returns nil if there are fewer nodes in the linked list than the index value
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def get_at_index(index)
+    return nil if index > (self.length - 1)
 
-    # method to print all the values in the linked list
-    # Time Complexity:  
-    # Space Complexity
-    def visit
-      raise NotImplementedError
+    current = @head
+    until index == 0
+      current = current.next
+      index -= 1
     end
 
-    # method to delete the first node found with specified value
-    # Time Complexity:  
-    # Space Complexity
-    def delete(value)
-      raise NotImplementedError
-    end
+    return current.data
+  end
 
-    # method to reverse the singly linked list
-    # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity:  
-    # Space Complexity
-    def reverse
-      raise NotImplementedError
+  # method to print all the values in the linked list
+  # Time Complexity:
+  # Space Complexity
+  def visit
+    raise NotImplementedError
+  end
+
+  # method to delete the first node found with specified value
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def delete(value)
+    return 0 if @head.nil?
+    current = @head
+    previous = nil
+
+    # if deleting from front
+    if current.data == value
+      @head = current.next
+      return
+    else # if deleting from somewhere in middle or end
+      until current.data == value
+        previous = current
+        current = current.next

```

Comment:
What about if the value is **not** in the list.  You could try to do `.data` on `nil`.